### PR TITLE
Extend Scorer

### PR DESCRIPTION
**Pull request by KEggensperger**
This PR enables ASKL to work with a wider range of metrics and scorers. Concretely, it does the following:

  * Allow a metric to require xdata 
  * Store xdata and pass it to the metric if required
  * Updates to the backend in automl-common to store and load xdata
  * Some fixed typos in the contribution guide

### CODE REVIEW

File: autosklearn/evaluation/train_evaluator.py [hunk 1]
```diff
@@ -654,6 +703,33 @@ def fit_predict_and_loss(self, iterative: bool = False) -> None:
                     if Y_optimization_pred[i] is not None
                 ]
             )
+            if isinstance(X_targets[0], np.ndarray):
```
Comment (mfeurer) [inline autosklearn/evaluation/train_evaluator.py:706]: There's now quite a bit of code duplicated, maybe it should go into its own function?

File: autosklearn/evaluation/train_evaluator.py [hunk 2]
```diff
@@ -400,6 +414,8 @@ def fit_predict_and_loss(self, iterative: bool = False) -> None:
                             if hasattr(self.Y_train, "iloc")
                             else self.Y_train[train_indices],
                             train_pred,
+                            # TODO: Check whether this is the correct xdata
```
Comment (mfeurer) [inline autosklearn/evaluation/train_evaluator.py:417]: How do we check this TODO?

**Review by mfeurer — COMMENTED** [review]
This looks good in general, but it would be great if this would be accompanied by some unit tests, for example for the train evaluator and the metrics module.

File: examples/40_advanced/example_metrics.py [hunk 3]
```diff
@@ -46,6 +46,12 @@ def error_wk(solution, prediction, extra_argument):
     return np.mean(solution != prediction)
 
 
+def metric_which_needs_x(solution, prediction, x_data, extra_argument):
+    # custom function defining accuracy
+    assert x_data is not None
+    return np.mean(solution[extra_argument, :] == prediction[extra_argument, :])
```
Comment (eddiebergman) [inline examples/40_advanced/example_metrics.py:52]: Can we illustrate with some arbitrary use of `x_data` for the example, i.e. make rows with NaNs count for twice as much?

**Review by eddiebergman — APPROVED** [review]

File: autosklearn/evaluation/train_evaluator.py (hunk 1 above)
Comment (KEggensperger) [reply autosklearn/evaluation/train_evaluator.py]: agreed and done

File: autosklearn/evaluation/train_evaluator.py (hunk 2 above)
Comment (KEggensperger) [reply autosklearn/evaluation/train_evaluator.py:437]: Sorry, I accidentally left this in the code. Removed.

File: examples/40_advanced/example_metrics.py (hunk 3 above)
Comment (KEggensperger) [reply examples/40_advanced/example_metrics.py]: I updated the example with a more meaningful metric

File: autosklearn/metrics/__init__.py [hunk 4]
```diff
@@ -104,19 +109,21 @@ def __call__(
         else:
             raise ValueError(type_true)
 
+        sc_args = {}  # type: Dict[str, Union[List[float], np.ndarray]]
```
Comment (mfeurer) [inline autosklearn/metrics/__init__.py:112]: This should be named `kwargs`

Comment (mfeurer) [reply autosklearn/metrics/__init__.py:112]: Also, what does the `sc` stand for?

File: autosklearn/metrics/__init__.py [hunk 5]
```diff
@@ -420,6 +446,8 @@ def calculate_scores(
     prediction: np.ndarray,
     task_type: int,
     metrics: Sequence[Scorer],
+    *,
+    x_data: Optional[SUPPORTED_XDATA_TYPES] = None,
```
Comment (mfeurer) [inline autosklearn/metrics/__init__.py:450]: This new argument should be added to the docstring.

File: autosklearn/metrics/__init__.py [hunk 6]
```diff
@@ -556,7 +597,11 @@ def calculate_losses(
 
 
 def compute_single_metric(
-    metric: Scorer, prediction: np.ndarray, solution: np.ndarray, task_type: int
+    metric: Scorer,
+    prediction: np.ndarray,
+    solution: np.ndarray,
+    task_type: int,
+    x_data: Optional[np.ndarray] = None,
```
Comment (mfeurer) [inline autosklearn/metrics/__init__.py:604]: Here as well.

File: autosklearn/metrics/__init__.py [hunk 7]
```diff
@@ -583,13 +628,18 @@ def compute_single_metric(
...
     return metric._sign * score
 
 
 def _compute_single_scorer(
-    metric: Scorer, prediction: np.ndarray, solution: np.ndarray, task_type: int
+    metric: Scorer,
+    prediction: np.ndarray,
+    solution: np.ndarray,
+    task_type: int,
+    x_data: Optional[SUPPORTED_XDATA_TYPES] = None,
```
Comment (mfeurer) [inline autosklearn/metrics/__init__.py:642]: And once more.

File: autosklearn/metrics/__init__.py [hunk 8]
```diff
@@ -32,11 +33,13 @@ def __init__(
         worst_possible_result: float,
         sign: float,
         kwargs: Any,
+        needs_x: bool = False,
```
Comment (mfeurer) [inline autosklearn/metrics/__init__.py:36]: I think we should use an upper-case X here to indicate that this is a 2d-array of samples and not a single sample.

File: examples/40_advanced/example_metrics.py [hunk 9]
```diff
@@ -186,3 +193,43 @@ def error_wk(solution, prediction, extra_argument):
...
+)
+cls.fit(X_train, y_train)
+
+predictions = cls.predict(X_test)
+score = metric_which_needs_x(
+    y_test,
+    predictions,
+    x_data=X_test,
+    consider_col=1,
+    threshold=18.8,
```
Comment (mfeurer) [inline examples/40_advanced/example_metrics.py:233]: Why do you change the threshold here?

**Review by mfeurer — COMMENTED** [review]
This looks all good, a new unit test in the file `test_train_evaluator.py`, which ensures that the data is correctly emitted to disk, would be great to have, though.

**Comment by eddiebergman** [comment]
I will run the benchmark on this, it's hard to implement this feature into the `ensemble_builder` until this is merged into dev.

File: autosklearn/metrics/__init__.py (hunk 4 above)
Comment (KEggensperger) [reply autosklearn/metrics/__init__.py:112]: Sure, changed it to scorer_kwargs which should be more descriptive

File: autosklearn/metrics/__init__.py (hunk 7 above)
Comment (KEggensperger) [reply autosklearn/metrics/__init__.py:642]: see above

File: autosklearn/metrics/__init__.py (hunk 8 above)
Comment (KEggensperger) [reply autosklearn/metrics/__init__.py]: done

File: examples/40_advanced/example_metrics.py (hunk 9 above)
Comment (KEggensperger) [reply examples/40_advanced/example_metrics.py]: sorry, this should be an extra argument to the metric. I renamed it

Comment (mfeurer) [reply examples/40_advanced/example_metrics.py]: Do you then also need to change the variable name in the definition of `accuracy_scorer` above:

```
accuracy_scorer = autosklearn.metrics.make_scorer(
    name="accu_X",
    score_func=metric_which_needs_x,
    optimum=1,
    greater_is_better=True,
    needs_proba=False,
    needs_X=True,
    needs_threshold=False,
    consider_col=1,
    threshold=20,
)
```

Comment (KEggensperger) [reply examples/40_advanced/example_metrics.py]: True. Done.

File: examples/40_advanced/example_metrics.py [hunk 10]
```diff
@@ -186,3 +193,43 @@ def error_wk(solution, prediction, extra_argument):
 predictions = cls.predict(X_test)
 score = error_rate(y_test, predictions)
 print(f"Error score {score:.3f} using {error_rate.name:s}")
+
+
+#############################################################################
+# Sixth example: Use a metric with additional argument which also needs xdata
+# ===============================================================
```
Comment (mfeurer) [inline examples/40_advanced/example_metrics.py:200]:
```suggestion
# ===========================================================================
```
This should fix the build.

**Review by mfeurer — APPROVED** [review]
